---
title: "Bioinfo Projekt"
author: "Ilya"
date: "5/1/2021"
output: html_document
---

```{r}
load("../2021-topic-01-team-03/dataset/prism_datasets.rda")
load("../2021-topic-01-team-03/dataset/cellline_datasets.rda")
```

#Guidelines

prism: effect of the treatment (columns) on cell growth of cell lines (rows).For example, BRD-A00077618-236-07-6::0.0024414::HTS002 indicates treatment with the drug BRD-A00077618-236-07-6, with a dose of 2.4 nM, in assay HTS002. Values represent log fold changes with respect to a control DMSO experiment.

prism.treat:information about the treatment (row) and the drug used

prism.cl: cell lines' info

prism.exp: gene TPM (transcripts per million) values. Indicate over/underexpression. Rows = cell lines, columns = genes

prism.cnv: gene copy number (CN) values. In absolute terms, CN = 2, since there are two alleles per genes. In cancer, genes might be amplified CN > 2 or deleted CN < 2

prism.snv: observed mutations in a sample. The isDeleterious flag specifies if mutation has a functional effect or not.

prism.achilles: gene KO scores, a measure of how essential/important is a particular gene for the cell survival

IDEA: use same indications as in PRISM paper to start filtering for relevant genes, cell lines? 

#Cleaning

##Selecting brain cancer specific cell lines in all data sets
```{r}
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names

names=brain_cancer_cl[,1] #saving the names as names

#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]
brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]
#brain_cancer_snv=prism.snv[names,] #does not make sense no cell line info, other type of filtering required
#brain_cancer_treat=prism.treat[names,] #does not make sense no cell line info, other type of filtering required
```

##NA removal from brain_cancer_achilles
```{r}
brain_cancer_achilles=na.omit(brain_cancer_achilles) #remove the entire cell lines from everywhere without KO scores
```

##NA for mean substituition
```{r}
for(i in 1:ncol(brain_cancer)){
  brain_cancer[is.na(brain_cancer[,i]), i] <- mean(brain_cancer[,i], na.rm = TRUE)
}

brain_cancer=brain_cancer[ , colSums(is.na(brain_cancer)) == 0]
```

##Brain_cancer distribution visualising
```{r}
max=apply(brain_cancer,2,max)
min=apply(brain_cancer,2,min)
mean=apply(brain_cancer,2,mean)
median=apply(brain_cancer,2,median)
#add quantiles to the dataframe?

drug_summary=data.frame(max, min, mean, median)
sum(is.na(drug_summary$mean))

hist(drug_summary$mean,main= "Distribution of means of each drug",xlab= "means of drugs", freq= F, breaks = 100)
abline(v = quantile(drug_summary$mean, c(0.25,0.5,0.75)), col = "blue", lwd = 2)
#what would be the type of this distribution? Poisson? Negative binominal?

median(drug_summary$mean) # 0.3 criteria from PRISM paper makes sense since it covers more than a half of the drugs
mean(drug_summary$mean) 

qqnorm(drug_summary$mean) #data not normally distributed
abline(a = 0, b = 0.33) #however one part of it seems to match the normal distribution, starting from -1 sample quantile

```

##Deviding drugs in doses groups
```{r}
dd1=grep("0.00061034", colnames(brain_cancer))
dd2=grep("0.0024414", colnames(brain_cancer))
dd3=grep("0.00976562", colnames(brain_cancer))
dd4=grep("0.0390625", colnames(brain_cancer))
dd5=grep("0.15625", colnames(brain_cancer))
dd6=grep("0.625", colnames(brain_cancer))
dd7=grep("10:", colnames(brain_cancer))
dd8=grep("2.5:", colnames(brain_cancer))
#positions of the respective doses in the dataframe 
doses=c(dd1,dd2,dd3,dd4,dd5,dd6,dd7,dd8)

d1=brain_cancer[,dd1]
d2=brain_cancer[,dd2]
d3=brain_cancer[,dd3]
d4=brain_cancer[,dd4]
d5=brain_cancer[,dd5]
d6=brain_cancer[,dd6]
d7=brain_cancer[,dd7]
d8=brain_cancer[,dd8]

hist(apply(d1,2,mean))#comparing means of 2 two different concentrations
hist(apply(d7,2,mean))

hist(apply(d1,2,mean), freq= F, breaks = 100)

#these are the standard doses summed up into seperate dataframes
sum(length(dd1), length(dd2), length(dd3), length(dd4),length(dd5), length(dd6),length(dd7),length(dd8)) #ups something missing

deviation=brain_cancer[,-doses] #lets see what we missed-> these are the doses that are not exact and have some kind of a deviation

questionable_drugs=prism.treat[colnames(deviation),]#we dont know if we can use these drugs, may be their doses deviate too far from the standard
```

#Working on drugs that do not fit the standart doses
```{r}
#rest of the doeses are all different. What do we do with them?

#lets keep from "deviation" only the ones that have a deviation less than "10%" from original

#to do: make a distribution test for a non normal distribution to justify the 10%
#filtering the doses that deviate less than 10%

keep_d1=which (0.00061034-0.1*0.00061034 < questionable_drugs[,2] & 0.00061034+0.1*0.00061034 >questionable_drugs[,2]) #apply the 10% gap and find which genes are that qualify for it
d1_extra=brain_cancer[,rownames(questionable_drugs[keep_d1,])] #create a dataframe with these genes 
d1=cbind(d1,d1_extra) #merge the obtained dataframe with the standard dose one

keep_d2=which (0.0024414-0.1*0.0024414 < questionable_drugs[,2] & 0.0024414+0.1*0.0024414 >questionable_drugs[,2])
d2_extra=brain_cancer[,rownames(questionable_drugs[keep_d2,])]
d2=cbind(d2,d2_extra)

keep_d3=which (0.00976562-0.1*0.00976562 < questionable_drugs[,2] & 0.00976562+0.1*0.00976562 >questionable_drugs[,2])
d3_extra=brain_cancer[,rownames(questionable_drugs[keep_d3,])]
d3=cbind(d3,d3_extra)

keep_d4=which (0.0390625-0.1*0.0390625 < questionable_drugs[,2] & 0.0390625+0.1*0.0390625 >questionable_drugs[,2])
d4_extra=brain_cancer[,rownames(questionable_drugs[keep_d4,])]
d4=cbind(d4,d4_extra)

keep_d5=which (0.15625-0.1*0.15625 < questionable_drugs[,2] & 0.15625+0.1*0.15625 >questionable_drugs[,2])
d5_extra=brain_cancer[,rownames(questionable_drugs[keep_d5,])]
d5=cbind(d5,d5_extra)

keep_d6=which (0.625-0.1*0.625 < questionable_drugs[,2] & 0.625+0.1*0.625 >questionable_drugs[,2])
d6_extra=brain_cancer[,rownames(questionable_drugs[keep_d6,])]
d6=cbind(d6,d6_extra)

keep_d7=which (10-0.1*10 < questionable_drugs[,2] & 10+0.1*10 >questionable_drugs[,2])
d7_extra=brain_cancer[,rownames(questionable_drugs[keep_d7,])]
d7=cbind(d7,d7_extra)

keep_d8=which (2.5-0.1*2.5 < questionable_drugs[,2] & 2.5+0.1*2.5 >questionable_drugs[,2])
d8_extra=brain_cancer[,rownames(questionable_drugs[keep_d8,])]
d8=cbind(d8,d8_extra)


#look at the drugs we did not keep
keep=c(keep_d1, keep_d2, keep_d3, keep_d4,keep_d5,keep_d6,keep_d7,keep_d8)
doses_outliers=deviation[,-keep]

mean(apply(deviation[,-keep],2,mean))#curioisty

```

##Selecting relevant drugs
```{r}
no_keep_drugs=colnames(deviation[,-keep])#vector with NAMES of drugs we dont want
no_keep_drugs=unique (grep(paste(no_keep_drugs,collapse="|"), colnames(brain_cancer)))#vector of POSITIONS of drugs we dont want in dataframe brain_cancer
brain_cancer=brain_cancer[,-no_keep_drugs]#removing the drugs we dont want
```

##Apllying the dosis separation of dataset brain_cancer on prism.treat
```{r}
brain_cancer_treat=prism.treat[colnames(brain_cancer),] #selecting only the drugs we previously filtered out from brain_cancer

dt1=brain_cancer_treat[colnames(d1),]
dt2=brain_cancer_treat[colnames(d2),]
dt3=brain_cancer_treat[colnames(d3),]
dt4=brain_cancer_treat[colnames(d4),]
dt5=brain_cancer_treat[colnames(d5),]
dt6=brain_cancer_treat[colnames(d6),]
dt7=brain_cancer_treat[colnames(d7),]
dt8=brain_cancer_treat[colnames(d8),]

#not sure this sorting makes sense now for the brain_cancer_treat, but now we have it in case we need it in the future

summary(brain_cancer_treat)#can be useful to identify trends or come up with some new ideas for filtering, interesting columns can be looked at separetely, for example:
summary(brain_cancer_treat$disease.area)

```

##Dividing cell lines into subgroups
```{r}
medulloblastoma=brain_cancer_cl$DepMap_ID[grep("Medulloblastoma", brain_cancer_cl$disease_subtype)]#we obtain the names of the cell lines that are classified as medulloblastoma

glioblastoma=brain_cancer_cl$DepMap_ID[grep("Glioblastoma", brain_cancer_cl$disease_subtype)]

glioma=brain_cancer_cl$DepMap_ID[grep("Glioma", brain_cancer_cl$disease_subtype)]

astrocytoma=brain_cancer_cl$DepMap_ID[grep("Astrocytoma", brain_cancer_cl$disease_subtype)]

disease_subtypes=c(medulloblastoma,glioblastoma,glioma,astrocytoma)
length(disease_subtypes)#making sure we didnt forget anything
```
##Clustering

```{r}

brain.km <- kmeans(scale(brain_cancer[, -5]), 6, nstart = 100)
# K-means clusters showing the group of each individuals
brain.km$cluster

```

##PCA for brain_cancer dataset
```{r}
library("factoextra")
dim(brain_cancer)
pca = prcomp(brain_cancer, center = TRUE, scale. = TRUE)

brain.pca <- prcomp(brain_cancer[, -5],  scale = TRUE)
# Coordinates of individuals
coordinates <- as.data.frame(get_pca_ind(brain.pca)$coord)
# Add clusters obtained using the K-means algorithm
coordinates$cluster <- factor(brain.km$cluster)
# Data inspection
head(coordinates)

```

## Analyse the variance of the eigenvalues
```{r}
# Percentage of variance explained by dimensions
eigenvalue <- round(get_eigenvalue(brain.pca), 1)
variance.percent <- eigenvalue$variance.percent
head(eigenvalue)
```


```{r}
library(ggpubr)
ggscatter(
  coordinates, x = "Dim.1", y = "Dim.2",
  color = "cluster", palette = "npg", ellipse = TRUE, ellipse.type = "convex",
  size = 1.5,
  legend = "right",
  ggtheme = theme_bw(),
  xlab = paste0("Dim 1 (", variance.percent[1], "% )" ),
  ylab = paste0("Dim 2 (", variance.percent[2], "% )" )
) +
  stat_mean(aes(color = cluster), size = 3)

```
```{r}

# Compute and plot wss for k = 2 to k = 12
k.max <- 12
data <- coordinates
wss <- sapply(1:k.max,
              function(k){kmeans(data, k, nstart=50,iter.max = 12 )$tot.withinss})
wss
plot(1:k.max, wss,
     type="b", pch = 19, frame = FALSE,
     xlab="Number of clusters K",
     ylab="Total within-clusters sum of squares")
```




