---
title: "Bioinfo Projekt"
author: "Ilya"
date: "5/1/2021"
output: html_document
---

```{r}
load("../2021-topic-01-team-03/dataset/prism_datasets.rda")
load("../2021-topic-01-team-03/dataset/cellline_datasets.rda")
```

#Guidelines

prism: effect of the treatment (columns) on cell growth of cell lines (rows).For example, BRD-A00077618-236-07-6::0.0024414::HTS002 indicates treatment with the drug BRD-A00077618-236-07-6, with a dose of 2.4 nM, in assay HTS002. Values represent log fold changes with respect to a control DMSO experiment.

prism.treat:information about the treatment (row) and the drug used

prism.cl: cell lines' info

prism.exp: gene TPM (transcripts per million) values. Indicate over/underexpression. Rows = cell lines, columns = genes

prism.cnv: gene copy number (CN) values. In absolute terms, CN = 2, since there are two alleles per genes. In cancer, genes might be amplified CN > 2 or deleted CN < 2

prism.snv: observed mutations in a sample. The isDeleterious flag specifies if mutation has a functional effect or not.

prism.achilles: gene KO scores, a measure of how essential/important is a particular gene for the cell survival

IDEA: use same indications as in PRISM paper to start filtering for relevant genes, cell lines? 

#Cleaning

##Selecting brain cancer specific cell lines in all data sets
```{r}
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names

names=brain_cancer_cl[,1] #saving the names as names

#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]
brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]
#brain_cancer_snv=prism.snv[names,] #does not make sense no cell line info, other type of filtering required
#brain_cancer_treat=prism.treat[names,] #does not make sense no cell line info, other type of filtering required
```

##NA removal
```{r}
#brain_cancer_achilles=na.omit(brain_cancer_achilles) #remove the entire cell lines from everywhere without KO scores?




```

##NA for mean substituition
```{r}
for(i in 1:ncol(brain_cancer)){
  brain_cancer[is.na(brain_cancer[,i]), i] <- mean(brain_cancer[,i], na.rm = TRUE)
}

brain_cancer=brain_cancer[ , colSums(is.na(brain_cancer)) == 0]
```

##Deviding drugs in doses groups
```{r}
d1=brain_cancer[,grep("0.00061034", colnames(brain_cancer))]
d2=brain_cancer[,grep("0.0024414", colnames(brain_cancer))]
d3=brain_cancer[,grep("0.00976562", colnames(brain_cancer))]
d4=brain_cancer[,grep("0.0390625", colnames(brain_cancer))]
d5=brain_cancer[,grep("0.15625", colnames(brain_cancer))]
d6=brain_cancer[,grep("0.625", colnames(brain_cancer))]
d7=brain_cancer[,grep("10:", colnames(brain_cancer))]
d8=brain_cancer[,grep("2.5:", colnames(brain_cancer))]
#these are the standard doses summed up into seperate dataframes

sum(length(d1), length(d2), length(d3), length(d4),length(d5), length(d6),length(d7),length(d8)) #ups something missing

deviation=brain_cancer[,-doses] #lets see what we missed-> these are the doses that are not exact and have some kind of a deviation

#rest of the doeses are all different. What do we do with them?

#lets keep from "deviation" only the ones that have a deviation less than 10% from original
questionable_drugs=prism.treat[colnames(deviation),]#we dont know if we can use these drugs, may be their doses deviate too far from the standard

#filtering the doses that deviate less than 10%

keep_d1=which (0.00061034-0.1*0.00061034 < questionable_drugs[,2] & 0.00061034+0.1*0.00061034 >questionable_drugs[,2]) #apply the 10% gap and find which genes are that qualify for it
d1_extra=brain_cancer[,rownames(questionable_drugs[keep_d1,])] #create a dataframe with these genes 
d1=merge(d1,d1_extra) #merge the obtained dataframe with the standard dose one

keep_d2=which (0.0024414-0.1*0.0024414 < questionable_drugs[,2] & 0.0024414+0.1*0.0024414 >questionable_drugs[,2])
d2_extra=brain_cancer[,rownames(questionable_drugs[keep_d2,])]
d2=merge(d2,d2_extra)

keep_d3=which (0.00976562-0.1*0.00976562 < questionable_drugs[,2] & 0.00976562+0.1*0.00976562 >questionable_drugs[,2])
d3_extra=brain_cancer[,rownames(questionable_drugs[keep_d3,])]
d3=merge(d3,d3_extra)

keep_d4=which (0.0390625-0.1*0.0390625 < questionable_drugs[,2] & 0.0390625+0.1*0.0390625 >questionable_drugs[,2])
d4_extra=brain_cancer[,rownames(questionable_drugs[keep_d4,])]
d4=merge(d4,d4_extra)

keep_d5=which (0.15625-0.1*0.15625 < questionable_drugs[,2] & 0.15625+0.1*0.15625 >questionable_drugs[,2])
d5_extra=brain_cancer[,rownames(questionable_drugs[keep_d5,])]
d5=merge(d5,d5_extra)

keep_d6=which (0.625-0.1*0.625 < questionable_drugs[,2] & 0.625+0.1*0.625 >questionable_drugs[,2])
d6_extra=brain_cancer[,rownames(questionable_drugs[keep_d6,])]
d6=merge(d6,d6_extra)

keep_d7=which (10-0.1*10 < questionable_drugs[,2] & 10+0.1*10 >questionable_drugs[,2])
d7_extra=brain_cancer[,rownames(questionable_drugs[keep_d7,])]
d7=merge(d7,d7_extra)

keep_d8=which (2.5-0.1*2.5 < questionable_drugs[,2] & 2.5+0.1*2.5 >questionable_drugs[,2])
d8_extra=brain_cancer[,rownames(questionable_drugs[keep_d8,])]
d8=merge(d8,d8_extra)


#look at the genes we did not keep
keep=c(keep_d1, keep_d2, keep_d3, keep_d4,keep_d5,keep_d6,keep_d7,keep_d8)
View(deviation[,-keep])

mean(apply(deviation[,-keep],2,mean))#curioisty


```

##Selecting relevant drugs
```{r}
brain_cancer[,drugs_not_keep] #dataframe with drugs that we dont wanna keep. find a way to remove them from dataframe brain_cancer

max=apply(brain_cancer,2,max)
min=apply(brain_cancer,2,min)
mean=apply(brain_cancer,2,mean)
median=apply(brain_cancer,2,median)
#add quantiles to the dataframe?

drug_summary=data.frame(max, min, mean, median)
sum(is.na(drug_summary$mean))


hist(drug_summary$mean, breaks = 100,)
abline(v = quantile(drug_summary$mean, c(0.25,0.5,0.75)), col = "blue", lwd = 2)
#what would be the type of this distribution? Poisson? Negative binominal?

median(drug_summary$mean) # 0.3 criteria from PRISM paper makes sense since it covers more than a half of the drugs
mean(drug_summary$mean) 

```

